heaps/min_heap.py

```diff
@@ -57,18 +69,46 @@ def heap_up(self, index):
             property is reestablished.
             
             This could be **very** helpful for the add method.
-            Time complexity: ?
-            Space complexity: ?
+            Time complexity: O(logn)
+            Space complexity: O(logn) because of the call stack
         """
-        pass
+        if index == 0:
+            return
+        parent = (index-1)//2
+        if self.store[parent].key > self.store[index].key:
+            self.swap(parent, index)
+            self.heap_up(parent)
+
 
     def heap_down(self, index):
         """ This helper method takes an index and 
             moves the corresponding element down the heap if it's 
             larger than either of its children and continues until
             the heap property is reestablished.
         """
-        pass
+        def get_valid_index_or_none(index):
+            if index < len(self.store):
+                return index
+            else:
+                return None
+
+        def get_key_or_none(index):
+            if not index:
+                return None
+            return self.store[index].key
+
+        left_index = get_valid_index_or_none(index * 2 + 1)
+        right_index = get_valid_index_or_none(index * 2 + 2)
+
+        left_key = get_key_or_none(left_index)
+        right_key = get_key_or_none(right_index)
+
+        index_to_swap = min((node for node in [(left_index, left_key), (right_index, right_key)] if node[1] is not None ), key=operator.itemgetter(1), default=None) 
```

Comment:
This is a really concise set of steps to filter out the invalid children, then select the minimum! However, it's a bit dense, which makes it a little tough to understand. Consider adding a comment about what it does, break it up into several lines with the intermediate products given descriptive names, or move into a helper function with a descriptive name.

An example of splitting across multiple lines
```py
        candidate_children = [(left_index, left_key), (right_index, right_key)]  # we could use a tuple rather than a list
        valid_children = (node for node in candidate_children if node[1] is not None)  # this actually makes a generator function, not a tuple
        index_to_swap = min(valid_children, key=operator.itemgetter(1), default=None) 
```